core-to-plc/src/Language/Plutus/CoreToPLC.hs

```diff
@@ -339,15 +338,54 @@ This is somewhat wasteful, since we may convert the expression twice, but it's d
 it's hard to tell if a GHC core expression will be a PLC value or not. Easiest to just try it.
 -}
 
+{- Note [Ordering of constructors]
+It is very important that we convert types and constructors consistently, especially between
+lifting at runtime and compilation via the plugin. The main place we can get bitten is ordering.
+
+GHC is under no obligation to give us any particular ordering of constructors, so we impose
```

Comment:
Is that true? I'd say, GHC needs to preserve the order of constructors as presented in the source; otherwise, `deriving Enum` is going to go quite wrong.

Reply:
From the [Generics documentation](http://hackage.haskell.org/package/base-4.12.0.0/docs/GHC-Generics.html#g:9):

> As `:+:` and `:*:` are just binary operators, one might ask what happens if the datatype has more than two constructors, or a constructor with more than two fields. The answer is simple: the operators are used several times, to combine all the constructors and fields as needed. However, users /should not rely on a specific nesting strategy/ for `:+:` and `:*:` being used. The compiler is free to choose any nesting it prefers.

But indeed this is about *nesting* not ordering. Still, I'd prefer not to make any assumptions since I cannot see anywhere that the order *is* guaranteed. Even if it is guaranteed in TH or Core, it is unclear whether it is guaranteed in Generics. Simpler just to impose our own order.